src/librustc_target/abi/mod.rs

```diff
@@ -871,8 +871,26 @@ impl Niche {
 
 #[derive(PartialEq, Eq, Hash, Debug, HashStable_Generic)]
 pub struct LayoutDetails {
-    pub variants: Variants,
+    /// Says where the fields are located.
+    /// Primitives and fieldless enums appear as unions without fields.
     pub fields: FieldPlacement,
+
+    /// Encodes information about multi-variant layouts.
+    /// Even with `Multiple` variants, a layout can still have fields! Those are then
+    /// shared between all variants. One of them will be the discriminant,
+    /// but e.g. generators can have more.
+    ///
+    /// A layout-guided recursive descent must first look at all the fields,
+    /// and only then check if this is a multi-variant layout and if so, proceed
+    /// with the active variant.
+    pub variants: Variants,
+
+    /// The `abi` defines how this data is passed between functions, and it defines
+    /// value restrictions via `valid_range`.
```

Comment:
Not just between functions, but it can also be taken advantage of for copying/moving values around inside a function (I understand if you don't want to include this).

Reply:
I guess a better way to phrase it is that this is the "by-value ABI", i.e. the ABI for manipulating the value without it having to be in memory.
Although having forced that explanation to work, maybe it's not better.

Reply:
I don't understand what you mean. I don't think there is anything within a function that we are allowed to do with e.g. a `ScalarPair` that we wouldn't also be allowed to do with a 3-field `Aggregate`.

Reply:
It's not about being *allowed*, it's about the *preferred* way to work with values.

Similar to how we track the "preferred alignment" vs "ABI(-mandated) alignment - nothing enforces it, but it's (potentially) good for performance to use the "preferred" one.

Reply:
But that is a separate thing then on top of what I am saying here, because when passed across function boundaries a `ScalarPair` *must* be treated in a certain way, or else the other side won't get the arguments correctly. Similarly, `valid_range` is a "must" clause. What you are saying sounds more like a "should" clause.

Reply:
> What you are saying sounds more like a "should" clause.

Not even, more like "can" - it's hard to do this without precomputation, which is why this exists in the first place.

> when passed across function boundaries a `ScalarPair` *must* be treated in a certain way, or else the other side won't get the arguments correctly

While this is irrelevant for FFI (`ScalarPair` is treated as any other aggregate there and not relied upon, IIRC), you are correct for the Rust ABI, and I guess your approach to this is fine.

I guess I wanted to have the other usecases of the `abi` field represented, but we don't have to.